### PR TITLE
Added forceban and forceunban, user info improvements, and small fixes

### LINKED ISSUE #154: [Proposal] Extend !info to show Infraction Counts
Current Info output:
image

I'd like to propose an increase in information to show infraction count(s)
```
> Infractions
Given: 999
Received: 0
```

### PR DESCRIPTION
**Pull request by jmazouri**
**Added EphemeralUser**
- Mocked user that implements IGuildUser, can set id, guild, and username
- Needed to facilitate forceban/forceunban and other features that require a user to be tracked who isn't a guild member

**MessageLogBehavior**
- Message hyperlink is now embedded via markdown, as the Discord client has been updated to support this

**ModerationModule**
- Added "forceban" and "forceunban" commands
- Added aliases to "ban" and "unban" respectively, with higher priority so they don't try to resolve users
- Added "OK" emoji reaction on successful command (as per a suggestion in the server)

**InfractionModule**
- Added Search command override to search by raw ID

**ModerationService**
- CreateInfractionAsync - we now try to retrieve the user from the guild, and if it doesn't exist, we create and track an EphemeralUser instance with mocked data
- DoRescindInfractionAsync - we now try to retrieve the user, and if we couldn't, only throw an error if we then try to unmute the user as that requires a proper instance. RemoveBanAsync is now passed the user id from the infraction

**UserInfoModule**
- Massive refactor (inline diff is useless)
  - Reduced code duplication
  - Modularized the sections a bit
  - Removed list formatting code as we already have Humanizer
- Made it work for users that aren't in the guild, but are tracked
  - Member info is not shown for such users
- Added infraction information (only shown if the user that runs the command has the ModerationRead claim, properly resolves #154 )

### CODE REVIEW

File: Modix.Data/Models/EphemeralUser.cs [hunk 1]
```diff
@@ -0,0 +1,111 @@
+﻿using Discord;
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Modix.Data.Models
+{
```
Comment (JakenVeina) [inline Modix.Data/Models/EphemeralUser.cs]: Would this fit better in Modix.Services? I don't like the idea of having Modix.Data have dependencies to Discord. And really, it only exists to simplify the service layer, so IGuildUser can be passed around in more scenarios, but the service layer should translate it all over to GuildUserEntity etc. to interact with Modix.Data.

File: Modix.Services/Moderation/ModerationService.cs [hunk 2]
```diff
@@ -246,7 +246,17 @@ public async Task CreateInfractionAsync(InfractionType type, ulong subjectId, st
             AuthorizationService.RequireClaims(_createInfractionClaimsByType[type]);
 
             var guild = await DiscordClient.GetGuildAsync(AuthorizationService.CurrentGuildId.Value);
-            var subject = await UserService.GetGuildUserAsync(guild.Id, subjectId);
+
+            IGuildUser subject;
+            try
```
Comment (JakenVeina) [inline Modix.Services/Moderation/ModerationService.cs]: Booooooo for exception-based flow-control. Is there no way to check whether a user exists, before attempting to Get() it?

File: Modix.Services/Moderation/ModerationService.cs [hunk 3]
```diff
@@ -538,17 +548,32 @@ private async Task DoRescindInfractionAsync(InfractionSummary infraction)
             await InfractionRepository.TryRescindAsync(infraction.Id, AuthorizationService.CurrentUserId.Value);
 
             var guild = await DiscordClient.GetGuildAsync(infraction.GuildId);
-            var subject = await UserService.GetGuildUserAsync(guild.Id, infraction.Subject.Id);
+
+            IGuildUser subject;
+            try
```
Comment (JakenVeina) [inline Modix.Services/Moderation/ModerationService.cs]: Same as above.

File: Modix/Modules/InfractionModule.cs [hunk 4]
```diff
@@ -67,39 +70,51 @@ public InfractionModule(IModerationService moderationService)
...
             var infractions = await ModerationService.SearchInfractionsAsync(
                 new InfractionSearchCriteria
                 {
-                    GuildId = subject.GuildId,
-                    SubjectId = subject.Id,
-                    IsDeleted = false
+                    GuildId = Context.Guild.Id,
+                    SubjectId = subjectId,
+                    IsDeleted = false,
+                    IsRescinded = false
```
Comment (JakenVeina) [inline Modix/Modules/InfractionModule.cs]: Why add `IsRescinded` here? Last I recall, we explicitly wanted to not filter out rescinded infractions. Like, if someone was tempmuted, for an hour, once that hour is up, Modix automatically rescinds the infraction, so it'll forevermore not show up in this command, when it's probably quite relevant to know if someone did something worthy of a tempmute.

File: Modix/Modules/ModerationModule.cs [hunk 5]
```diff
@@ -18,39 +18,53 @@ public ModerationModule(IModerationService moderationService)
             ModerationService = moderationService;
         }
 
+        public async Task AddConfirmationEmoji()
```
Comment (JakenVeina) [inline Modix/Modules/ModerationModule.cs]: I would just make this `AddConfirmation` or `ReplyConfirmation` as it's not NECESSARILY gonna be an Emoji. ;)

File: Modix/Modules/ModerationModule.cs [hunk 6]
```diff
@@ -63,32 +77,68 @@ public ModerationModule(IModerationService moderationService)
...
+            await AddConfirmationEmoji();
+        }
+
+        [Command("forceban")]
+        [Alias("ban")]
+        [Summary("Ban a user from the guild, even if they are not a member.")]
+        [Priority(10)]
+        public async Task Forceban(
+            [Summary("The id of the user to be banned.")]
+                ulong subject,
```
Comment (JakenVeina) [inline Modix/Modules/ModerationModule.cs]: Should be `subjectId`.

File: Modix/Modules/ModerationModule.cs [hunk 7]
```diff
@@ -63,32 +77,68 @@ public ModerationModule(IModerationService moderationService)
...
+            await AddConfirmationEmoji();
+        }
+
+        [Command("forceunban")]
+        [Alias("unban")]
+        [Summary("Remove a ban that has been applied to a user.")]
+        [Priority(10)]
+        public async Task ForceUnban(
+            [Summary("The id of the user to be un-banned.")]
+                ulong subject)
```
Comment (JakenVeina) [inline Modix/Modules/ModerationModule.cs]: Should be `subjectId`.

File: Modix/Modules/UserInfoModule.cs [hunk 8]
```diff
@@ -9,158 +9,170 @@
...
+                {
+                    new SortingCriteria
+                    {
+                        PropertyName = "CreateAction.Created",
+                        Direction = SortDirection.Descending
+                    }
+                }
+            );
+
+            var noticeCount = infractions.Count(x => x.Type == InfractionType.Notice);
```
Comment (JakenVeina) [inline Modix/Modules/UserInfoModule.cs]: You should definitely write a new repository method for this, that just does the counts in the query. I did something like this in `PromotionCampaignRepository` where it counts up the total number of promotion comments, by sentiment, and returns an `IDictionary<PromotionSentiment, int>`.

File: Modix.Services/Moderation/ModerationService.cs (hunk 2 above)
Comment (jmazouri) [reply Modix.Services/Moderation/ModerationService.cs]: 100% agreed - but, not without refactoring `UserService`, unfortunately. I'll look into that, maybe a `TryGetUser` or similar.

File: Modix.Data/Models/EphemeralUser.cs (hunk 1 above)
Comment (jmazouri) [reply Modix.Data/Models/EphemeralUser.cs]: Hmm, a good point. I'll move it.

File: Modix/Modules/InfractionModule.cs (hunk 4 above)
Comment (jmazouri) [reply Modix/Modules/InfractionModule.cs]: What I probably should do is modify the command embed here - right now, there's no indicator of if an infraction is active or rescinded, so I didn't want to cause confusion. I'll do that instead.

Comment (JakenVeina) [reply Modix/Modules/InfractionModule.cs]: If it's in the interest of saving space, I suppose I'm okay with leaving them out.

I'm tempted to just say "forget displaying infractions in Discord, just link to the web and be done with it."

**Pull request merged**